### PR TITLE
BSC-10447 structuring the eeprom data

### PR DESCRIPTION
**Pull request by amospaul-bsn**
Reviewer: @3mp4y5 
Initial version.

### CODE REVIEW

File: openflow_input/bsn_port_desc_module_eeprom [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+    OFP_BSN_UNIT_KM = 0x4,
+    OFP_BSN_UNIT_M = 0x5,
+    OFP_BSN_UNIT_S = 0x6,
+    OFP_BSN_UNIT_C = 0x7,
+    OFP_BSN_UNIT_F = 0x8
+};
+
+struct of_bsn_unit {
+    uint32_t value;
+    /* We could probably use an enum to convey the unit */
```
Comment (kenchiang) [inline openflow_input/bsn_port_desc_module_eeprom]: suggest updating this placeholder comment

File: openflow_input/bsn_port_desc_module_eeprom [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+    enum ofp_bsn_unit_type unit;
+};
+
+enum ofp_bsn_module_eeprom_recv_signal_type(wire_type=uint8_t, bitmask=False) {
+    OFP_BSN_MODULE_EEPROM_RECV_SIGNAL_OMA = 0x0,
+    OFP_BSN_MODULE_EEPROM_RECV_SIGNAL_AVG = 0x1
```
Comment (kenchiang) [inline openflow_input/bsn_port_desc_module_eeprom]: let's add some explanatory comment describing each of the non-obvious values ; for instance, what is "OMA" ?

File: openflow_input/bsn_port_desc_module_eeprom [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+};
+
+enum ofp_bsn_module_eeprom_diag_type(wire_type=uint8_t, bitmask=False) {
+    OFP_BSN_MODULE_EEPROM_DIAG_CURR = 0x0,
+    OFP_BSN_MODULE_EEPROM_DIAG_VOLT = 0x1,
+    OFP_BSN_MODULE_EEPROM_DIAG_TEMP = 0x2,
+    OFP_BSN_MODULE_EEPROM_DIAG_RECP = 0x3,
+    OFP_BSN_MODULE_EEPROM_DIAG_OUTP = 0x4
```
Comment (kenchiang) [inline openflow_input/bsn_port_desc_module_eeprom]: let's add some explanatory comment describing each of the non-obvious values

File: openflow_input/bsn_port_desc_module_eeprom [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+    OFP_BSN_MODULE_EEPROM_IDENTIFIER_NONE = 0x0,
+    OFP_BSN_MODULE_EEPROM_IDENTIFIER_GBIC = 0x1,
+    OFP_BSN_MODULE_EEPROM_IDENTIFIER_ONBOARD = 0x2,
+    OFP_BSN_MODULE_EEPROM_IDENTIFIER_SFP = 0x3
+};
+
+enum ofp_bsn_module_eeprom_extended_identifier(wire_type=uint8_t, bitmask=False) {
+    OFP_BSN_MODULE_EEPROM_EXT_IDENTIFIER_NON_GBIC = 0x00,
+    OFP_BSN_MODULE_EEPROM_EXT_IDENTIFIER_GBIC_SFP = 0x04,
+    OFP_BSN_MODULE_EEPROM_EXT_IDENTIFIER_GBIC_MOD_DEF = 0x07
```
Comment (kenchiang) [inline openflow_input/bsn_port_desc_module_eeprom]: are these values specified elsewhere, and you are just enumerating only the ones you expect to see ?

File: openflow_input/bsn_port_desc_module_eeprom [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+};
+
+enum ofp_bsn_module_eeprom_connector(wire_type=uint8_t, bitmask=False) {
+    OFP_BSN_MODULE_EEPROM_CONNECTOR_NONE = 0x00,
+    OFP_BSN_MODULE_EEPROM_CONNECTOR_SC = 0x01,
+    OFP_BSN_MODULE_EEPROM_CONNECTOR_FC_1_COPPER = 0x02,
+    OFP_BSN_MODULE_EEPROM_CONNECTOR_FC_2_COPPER = 0x03,
+    OFP_BSN_MODULE_EEPROM_CONNECTOR_BNC_TNC = 0x04,
+    OFP_BSN_MODULE_EEPROM_CONNECTOR_FC_COAX = 0x05,
+    OFP_BSN_MODULE_EEPROM_CONNECTOR_FIBERE_JACK = 0x06,
```
Comment (kenchiang) [inline openflow_input/bsn_port_desc_module_eeprom]: typo ?

File: openflow_input/bsn_port_desc_module_eeprom [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+    OFP_BSN_MODULE_EEPROM_CONNECTOR_LC = 0x07,
+    OFP_BSN_MODULE_EEPROM_CONNECTOR_MT_RJ = 0x08,
+    OFP_BSN_MODULE_EEPROM_CONNECTOR_MU = 0x09,
+    OFP_BSN_MODULE_EEPROM_CONNECTOR_SG = 0x0a,
+    OFP_BSN_MODULE_EEPROM_CONNECTOR_OPTICAL_PIGTAIL = 0x0b,
+    OFP_BSN_MODULE_EEPROM_CONNECTOR_MPO_PARALLEL_OPTIC = 0x0c,
+    OFP_BSN_MODULE_EEPROM_CONNECTOR_HSSDC = 0x20,
+    OFP_BSN_MODULE_EEPROM_CONNECTOR_COPPER_PIGTAIL = 0x21,
+    OFP_BSN_MODULE_EEPROM_CONNECTOR_RJ45 = 0x22
```
Comment (kenchiang) [inline openflow_input/bsn_port_desc_module_eeprom]: are these values specified elsewhere, and you are just enumerating only the ones you expect to see ?

File: openflow_input/bsn_port_desc_module_eeprom [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+    OFP_BSN_MODULE_EEPROM_ALARM_LOW_SET = 0x02,
+    OFP_BSN_MODULE_EEPROM_ALARM_HI_WARN_SET = 0x04,
+    OFP_BSN_MODULE_EEPROM_ALARM_LOW_WARN_SET = 0x08,
+};
+
+struct of_port_desc_prop_bsn_alarm : of_port_desc_prop_bsn {
+    uint16_t    type == 0xffff;
+    uint16_t    length;
+    uint32_t    experimenter == 0x5c16c7;
+    uint32_t    exp_type == 8;
```
Comment (kenchiang) [inline openflow_input/bsn_port_desc_module_eeprom]: it looks like you are extending the of_port_desc_prop_bsn with more types, but there are no new ones defined on the extensions wiki page.  please update the wiki page with the ones you are adding.
also there does not appear to be an of_port_desc_prop_bsn with exp_type == 9 ; is this intentional ?

File: openflow_input/bsn_port_desc_module_eeprom [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+    enum ofp_bsn_module_eeprom_alarm_set alarm_set;
+    uint32_t high;
+    uint32_t high_warn;
+    uint32_t low;
+    uint32_t low_warn;
+    enum ofp_bsn_module_eeprom_diag_type alarm_type;
+    /* We could probably use an enum to convey the unit */
```
Comment (kenchiang) [inline openflow_input/bsn_port_desc_module_eeprom]: suggest updating placeholder comment

File: openflow_input/bsn_port_desc_module_eeprom [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+    enum ofp_bsn_unit_type unit;
+};
+
+struct ofp_bsn_module_eeprom_transceiver {
+    uint64_t    codes; /* The codes are in 8 bytes */
+    //list(of_port_desc_prop_t) info;
```
Comment (kenchiang) [inline openflow_input/bsn_port_desc_module_eeprom]: not sure what's going on here ; can you clarify ?  is this for future expansion ?  or a placeholder ?

File: openflow_input/bsn_port_desc_module_eeprom [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+    uint32_t high;
+    uint32_t high_warn;
+    uint32_t low;
+    uint32_t low_warn;
+    enum ofp_bsn_module_eeprom_diag_type alarm_type;
+    /* We could probably use an enum to convey the unit */
+    enum ofp_bsn_unit_type unit;
+};
+
+struct ofp_bsn_module_eeprom_transceiver {
```
Comment (kenchiang) [inline openflow_input/bsn_port_desc_module_eeprom]: this looks like it should be derived from of_port_desc_prop_bsn.  for example, see of_port_desc_prop_bsn_alarm above.

File: openflow_input/bsn_port_desc_module_eeprom [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+};
+*/
+
+struct of_port_desc_prop_bsn_ethtool : of_port_desc_prop_bsn {
+    uint16_t    type == 0xffff;
+    uint16_t    length;
+    uint32_t    experimenter == 0x5c16c7;
+    uint32_t    exp_type == 6;
+    uint16_t    key_length == length(compliance);
+    /* Available propertties */
```
Comment (kenchiang) [inline openflow_input/bsn_port_desc_module_eeprom]: typo

File: openflow_input/bsn_port_desc_module_eeprom [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+    of_octets_t data;
+};
+*/
+
+struct of_port_desc_prop_bsn_ethtool : of_port_desc_prop_bsn {
+    uint16_t    type == 0xffff;
+    uint16_t    length;
+    uint32_t    experimenter == 0x5c16c7;
+    uint32_t    exp_type == 6;
+    uint16_t    key_length == length(compliance);
```
Comment (kenchiang) [inline openflow_input/bsn_port_desc_module_eeprom]: what is the intent here?

File: openflow_input/bsn_port_desc_module_eeprom [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+    uint64_t vendor_name_lo; /* Vendor name as ASCII */
+    uint64_t vendor_name_hi; /* Vendor name as ASCII */
+    uint32_t vendor_oui; /* a list of 3 bytes */
+    uint64_t vendor_pn_lo; /* Vendor PN as ASCII */
+    uint64_t vendor_pn_hi; /* Vendor PN as ASCII */
+    uint32_t vendor_rev; /* Vendor revision as ASCII */
+
+    //list(of_port_desc_prop_t) compliance;
+    /* module eeprom can have opt diagnostics optionally */
+    list(of_port_desc_prop_t) more_properties;
```
Comment (kenchiang) [inline openflow_input/bsn_port_desc_module_eeprom]: maybe we should flesh this particular property out some more first ?

File: openflow_input/bsn_port_desc_module_eeprom (hunk 1 above)
Comment (amospaul-bsn) [reply openflow_input/bsn_port_desc_module_eeprom]: Ok sure.

File: openflow_input/bsn_port_desc_module_eeprom (hunk 2 above)
Comment (amospaul-bsn) [reply openflow_input/bsn_port_desc_module_eeprom]: Sure

File: openflow_input/bsn_port_desc_module_eeprom (hunk 3 above)
Comment (amospaul-bsn) [reply openflow_input/bsn_port_desc_module_eeprom]: will do

File: openflow_input/bsn_port_desc_module_eeprom (hunk 5 above)
Comment (amospaul-bsn) [reply openflow_input/bsn_port_desc_module_eeprom]: Yes will correct it.

File: openflow_input/bsn_port_desc_module_eeprom (hunk 4 above)
Comment (amospaul-bsn) [reply openflow_input/bsn_port_desc_module_eeprom]: These are the only possible values I expect to see.

File: openflow_input/bsn_port_desc_module_eeprom (hunk 7 above)
Comment (amospaul-bsn) [reply openflow_input/bsn_port_desc_module_eeprom]: Everything above 8 is commented out. Will need to remove them

File: openflow_input/bsn_port_desc_module_eeprom (hunk 8 above)
Comment (amospaul-bsn) [reply openflow_input/bsn_port_desc_module_eeprom]: yes will do

File: openflow_input/bsn_port_desc_module_eeprom (hunk 9 above)
Comment (amospaul-bsn) [reply openflow_input/bsn_port_desc_module_eeprom]: the commented code has to be removed. let me update it.

File: openflow_input/bsn_port_desc_module_eeprom [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+    uint16_t    type == 0xffff;
+    uint16_t    length;
+    uint32_t    experimenter == 0x5c16c7;
+    uint32_t    exp_type == 7;
+    uint32_t    laser_bias_curr;
+    uint32_t    laser_output_power;
+    enum ofp_bsn_module_eeprom_recv_signal_type laser_receiver_power_type;
+    uint32_t    laser_receiver_power;
+    uint32_t    module_temp;
+    uint32_t    module_voltage;
```
[the remark this answers is not in the document]
Comment (amospaul-bsn) [reply openflow_input/bsn_port_desc_module_eeprom]: will add them

**Pull request merged**